synapse/handlers/user_directory.py

```diff
@@ -217,30 +229,32 @@ async def _handle_deltas(self, deltas: List[Dict[str, Any]]) -> None:
                     else:
                         logger.debug("Server is still in room: %r", room_id)
 
-                is_support = await self.store.is_support_user(state_key)
-                if not is_support:
-                    if change is None:
-                        # Handle any profile changes
+                is_support_user = await self.store.is_support_user(state_key)
+                if is_support_user:
+                    continue
+
+                if joined is MatchChange.no_change:
+                    # Handle any profile changes for remote users
+                    if not self.is_mine_id(state_key):
                         await self._handle_profile_change(
                             state_key, room_id, prev_event_id, event_id
                         )
-                        continue
 
-                    if change:  # The user joined
-                        event = await self.store.get_event(event_id, allow_none=True)
-                        # It isn't expected for this event to not exist, but we
-                        # don't want the entire background process to break.
-                        if event is None:
-                            continue
+                elif joined is MatchChange.now_true:  # The user joined
+                    event = await self.store.get_event(event_id, allow_none=True)
+                    # It isn't expected for this event to not exist, but we
+                    # don't want the entire background process to break.
+                    if event is None:
+                        continue
 
-                        profile = ProfileInfo(
-                            avatar_url=event.content.get("avatar_url"),
-                            display_name=event.content.get("displayname"),
-                        )
+                    profile = ProfileInfo(
+                        avatar_url=event.content.get("avatar_url"),
+                        display_name=event.content.get("displayname"),
+                    )
 
-                        await self._handle_new_user(room_id, state_key, profile)
-                    else:  # The user left
-                        await self._handle_remove_user(room_id, state_key)
+                    await self._handle_new_user(room_id, state_key, profile)
```

Comment:
As you mentioned, this has the hole of 'new joins will still trust the event content'.

Since we're expecting local users to populate themselves into the directory as soon as they update their profile, there's no need to run the risk of overwriting that with the content from the event (which feasibly may allow per-room information in the future).

I think it might be worth splitting up `handle_new_user`, which is currently these two chunks, into two separate things:

* updating the profile itself (is that just `update_profile_in_user_dir`?)
  * **now**: don't trigger on state events in rooms for local users (but continue doing so for remote users until we get the federation requests sorted out)
  * **future**: only trigger on local user profile updates and when we get responses for our federation profile queries
* updating the 'users in public rooms' and 'users that share private rooms' tables — never from state events
  * carry on triggering when you notice a new join in any room

shout if this seems silly

Reply:
> As you mentioned, this has the hole of 'new joins will still trust the event content'.

Where does the event content actually come from? It seems to come from the homeserver of the person joining; can't see anything in the CSAPI spec suggesting clients can customise this.

> updating the 'users in public rooms' and 'users that share private rooms' tables — never from state events

Erm, what does state event mean here? I think you're getting at "we don't need to do this when someone's name/avatar changes", just on joining or leaving?

Reply:
> Where does the event content actually come from? It seems to come from the homeserver of the person joining; can't see anything in the CSAPI spec suggesting clients can customise this.

Yes, that's right. Maybe the CS API doesn't support it today, but it's not unreasonable to think it would be possible one day (and in our case, I think it'd be more consistent if we didn't use the nicknames from state events in some cases and not others). -- Or I can imagine your name being inherited from a space, when you join a room that's in that space. I'm aware some people are quite enthusiastic about using spaces for segregating their (work, play/gaming, personal, ...) 'lives' and I believe this kind of thing has already been requested (since people playing games put 'gamer tags' in their name or something, but I can also see it being useful for having your real name in some contexts and not others, for example).

> Erm, what does state event mean here? I think you're getting at "we don't need to do this when someone's name/avatar changes", just on joining or leaving?

I'm specifically talking about the `m.room.member` state event, which (if you know what state events are, if not ask!) form a key-value mapping {user_id → (left/joined/kicked/banned/knocking, display name, avatar)}.

Type `/devtools` in Element Web and press 'Explore room state', if that helps you see things.

I'm trying to get at: we shouldn't be using `m.room.member` state event displaynames in the user directory *ever* (except for remote users for now, until we can properly accommodate those).

Crucially, the fact that a state event corresponds to a new user joining a room or them 'just' updating their profile information, should not influence that we use/don't use that information.